controlplane/kubeadm/controllers/controller.go

```diff
@@ -254,31 +254,36 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, nil
 	}
 
-	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
-	requireUpgrade := controlPlane.MachinesNeedingUpgrade()
-	// Upgrade takes precedence over other operations
-	if len(requireUpgrade) > 0 {
+	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines, logger)
+	if controlPlane.HasOperationInProgress() {
+		return ctrl.Result{}, nil
+	}
+
+	if controlPlane.NeedsInitialization() {
+		logger.Info("Initializing control plane", "Desired", int(*kcp.Spec.Replicas), "Existing", len(ownedMachines))
+		return r.initializeControlPlane(ctx, cluster, kcp, controlPlane)
+	}
+
+	if controlPlane.NeedsUpgrade() {
```

Comment:
Do we need to worry about the existence of any failed machines that are not "owner remediated" prior to allowing any upgrade or scale up actions, it looks like they would be ignored by `controlPlane.HasOperationInProgress()`, but I suspect I may be missing some additional validation that is happening elsewhere.

Reply:
Machines with `FailureMessage`/`Failure but no failing healthcheck condition? You are correct there's nothing handling that afaik. What would we want to do?

Reply:
I opened #3230 for this